janus_core/helpers/struct_io.py

```diff
@@ -0,0 +1,291 @@
+    ----------
+    struct : Optional[MaybeSequence[Atoms]]
+        ASE Atoms structure(s) to simulate. Required if `struct_path` is None. Default
+        is None.
+    struct_path : Optional[PathLike]
+        Path of structure to simulate. Required if `struct` is None. Default is None.
+    read_kwargs : ASEReadArgs
+        Keyword arguments to pass to ase.io.read. Default is {}.
+    sequence_allowed : bool
+        Whether a sequence of Atoms objects is allowed. Default is True.
+    arch : Architectures
+        MLIP architecture to use for calculations. Default is "mace_mp".
+    device : Devices
+        Device to run model on. Default is "cpu".
+    model_path : Optional[PathLike]
+        Path to MLIP model. Default is `None`.
+    calc_kwargs : Optional[dict[str, Any]]
+        Keyword arguments to pass to the selected calculator. Default is {}.
+    set_calc : Optional[bool]
+        Whether to set (new) calculators for structures.  Default is True if
+        `struct_path` is specified or `struct` is missing calculators, else default is
+        False.
+    logger : Optional[logging.Logger]
+        Logger if log file has been specified. Default is None.
+
+    Returns
+    -------
+    MaybeSequence[Atoms]
+        Structure(s) with attached MLIP calculators.
+    """
+    read_kwargs = read_kwargs if read_kwargs else {}
+
+    # Validate parameters
+    if not struct and not struct_path:
+        raise ValueError(
+            "Please specify either the ASE Atoms structure (`struct`) "
+            "or a path to the structure file (`struct_path`)"
+        )
+
+    if struct and struct_path:
+        raise ValueError(
+            "You cannot specify both the ASE Atoms structure (`struct`) "
+            "and a path to the structure file (`struct_path`)"
+        )
+
+    # Read from file
+    if struct_path:
+        if logger:
+            logger.info("Reading structures from file.")
+        struct = read(struct_path, **read_kwargs)
+        if logger:
+            logger.info("Structures read from file.")
+
+        # Return single Atoms object if reading only one image
+        if len(struct) == 1:
+            struct = struct[0]
+
+    # Check struct is valid type
+    if not isinstance(struct, (Atoms, Sequence)) or isinstance(struct, str):
+        raise ValueError("`struct` must be an ASE Atoms object or sequence of Atoms")
+
+    # Check struct is valid length
+    if not sequence_allowed and isinstance(struct, Sequence):
+        raise NotImplementedError(
+            "Only one Atoms object at a time can be used for this calculation"
+        )
+
+    # If set_calc not specified, set to True if reading file or calculators not attached
+    if set_calc is None:
+        if struct_path:
+            set_calc = True
+
+        elif isinstance(struct, Atoms):
+            set_calc = struct.calc is None
+
+        elif isinstance(struct, Sequence):
+            set_calc = any(image.calc is None for image in struct)
```

Comment:
Optionally?
```python
set_calc = (struct_path or
            (isinstance(struct, Atoms) and struct.calc is None) or
            (isinstance(struct, Sequence) and any(image.calc is None for image in struct)))
```

Reply:
I think this is probably a bit less readable